0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+- RFC PR: [dafny-lang/rfcs#9](https://github.com/dafny-lang/rfcs/pull/9)
+- Dafny Issue: [dafny-lang/dafny#1753](https://github.com/dafny-lang/dafny/issues/1753)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes adding two closely-related new features to Dafny for working with collections :
+
+1. `foreach` loops over any enumerable domain, initially only supporting the builtin collection types but leaving the door open for user-defined collections.
+2. Generalized sequence comprehension expressions, similar to the existing `set` and `map` comprehension expressions and more flexible than the current `seq(5, i => i*i)` form.
+
+# Motivation
+[motivation]: #motivation
+
+Loops are notoriously difficult for Dafny users to verify, and coming up with the correct loop invariants to prove a loop correct 
+is particularly challenging. The recent addition of `for` loops was a good step to abstract away from raw `while` loops,
+but only supports consecutive integer indices. They still fall short of the expressiveness and economy of `for` or `foreach` loops in
+many mainstream programming languages, which generally iterate through the contents of a collection of some kind.
+It is a well-established best practice to avoid manual loop indexes wherever possible, as they force the assumption that datatypes
+support efficient random access, and are more likely to be used incorrectly by accident:
+
+```dafny
+// Before:
+method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
+  for i := 0 to |s| {
+    var left := s[i];
+    for j := 0 to |s| {
+      var right := s[i]; // Whoops!
+      result := result + (left, right);  
+    }
+  }
+}
+
+// After:
+method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
+  foreach left | left in s {
+    foreach right | right in s {
+      result := result + (left, right);
+    }
+  }
+}
+
+// Or even better, as an expression:
+function AllPairs(s: seq<nat>): seq<(nat, nat)> {
+  seq left, right | left in s && right in s
+}
+```
+
+A more serious shortcoming is that there is currently no efficient way to iterate over the elements of a `set` at all.
+The best alternative is to use the assign-such-that operator and set subtraction, illustrated in the reference manual as follows
+([sec 10.5.2](https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#1052-sets)):
+
+```dafny
+// s is a set<int>
+var ss := s;
+while ss != {}
+  decreases |ss|
+{
+  var i: int :| i in ss;
+  ss := ss - {i};
+  print i, "\n";
+}
+```
+
+This is functionally correct, but because set subtraction requires creating a fresh copy,
+when compiled this loop requires quadratic time and memory (or at least garbage-collection work) in the size of the set.
+The only way to avoid this cost is to write additional external code in the target language to interface directly with the
+internal runtime implementation of `set` values.
+
+By comparison, the equivalent `foreach` loop is:
+
+```dafny
+foreach i | i in s {
+  print i, "\n";
+}
```

Comment:
It's a strong future possibility at least. Only supporting these higher-level operations over the elements of a set buys us a ton for now though, so I'm comfortable leaving it as future work.

Reply:
I really want to be able to write those three lines, too!

I'd also love to see set iterators in the future, but agree that it's not necessary in the first iteration.

Reply:
In line with what @samuelgruetter said above, though: since in Dafny `i in [1,2,3] <==> i in [3,2,1]`, shouldn't `foreach i | i in [1,2,3] { … }` be the same as `foreach i | i in [3,2,1] { … }`?

Reply:
Since `foreach` is not an expression but a statement, its semantics require an ordering, and we would syntactically extract this ordering from the expression. In [my answer below](https://github.com/dafny-lang/rfcs/pull/9#issuecomment-1087668318), I explain how we could encode the ordering for the verifier, so that the two foreach statement you mention wouldn't be the same.

Reply:
That's not right: `var s: seq<int> :| set(s) == {1,2,3}` is an expression, and it doesn't require an ordering.

Reply:
It does not require ordering because there is only one statement to execute :-)
To be fair,  I should have said, _is a statement that will execute other statements in some orders_